jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp

```diff
@@ -1448,44 +1448,128 @@ ecma_builtin_array_prototype_object_sort (ecma_value_t this_arg, /**< this argum
+    }
   }
 
-  JERRY_ASSERT (copied_num == len || !ecma_is_completion_value_empty (ret_value));
+  JERRY_ASSERT (copied_num == defined_prop_count || !ecma_is_completion_value_empty (ret_value));
 
   /* Sorting. */
-  if (len > 1 && ecma_is_completion_value_empty (ret_value))
+  if (copied_num > 1 && ecma_is_completion_value_empty (ret_value))
   {
     ECMA_TRY_CATCH (sort_value,
                     ecma_builtin_array_prototype_object_array_heap_sort_helper (values_buffer,
-                                                                                (int)(len - 1),
+                                                                                (int)(copied_num - 1),
                                                                                 arg1),
                     ret_value);
     ECMA_FINALIZE (sort_value);
   }
 
-  if (ecma_is_completion_value_empty (ret_value))
+  /* Put sorted values to the front of the array. */
+  for (uint32_t index = 0; index < copied_num && ecma_is_completion_value_empty (ret_value); index++)
   {
-    /*
-     * FIXME: Casting len to ecma_length_t may overflow, but since ecma_length_t is still at least
-     * 16 bits long, with an array of that size, we would run out of memory way before this happens.
-     */
-    JERRY_ASSERT ((ecma_length_t) len == len);
-    /* Copy the sorted array into a new array. */
-    ret_value = ecma_op_create_array_object (values_buffer, (ecma_length_t) len, false);
+    ecma_string_t *index_string_p = ecma_new_ecma_string_from_uint32 (index);
+    ECMA_TRY_CATCH (put_value,
+                    ecma_op_object_put (obj_p, index_string_p, values_buffer[index], true),
+                    ret_value);
+    ECMA_FINALIZE (put_value);
+    ecma_deref_ecma_string (index_string_p);
+  }
+
+  /* Undefined properties should be in the back of the array. */
+  for (ecma_property_t *property_p = ecma_get_property_list (obj_p);
+       property_p != NULL && ecma_is_completion_value_empty (ret_value);
+       )
+  {
+    ecma_string_t *property_name_p;
+
+    if (property_p->type == ECMA_PROPERTY_NAMEDDATA)
+    {
+      property_name_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
+                                                   property_p->u.named_data_property.name_p);
+    }
+    else if (property_p->type == ECMA_PROPERTY_NAMEDACCESSOR)
+    {
+      property_name_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
+                                                   property_p->u.named_accessor_property.name_p);
+    }
+    else
+    {
+      continue;
+    }
+
+    ecma_property_t *next_prop = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p);
+
+    uint32_t index;
+    if (ecma_string_get_array_index (property_name_p, &index) && index < len && index >= copied_num)
+    {
+      ECMA_TRY_CATCH (del_value, ecma_op_object_delete (obj_p, property_name_p, true), ret_value);
+      ECMA_FINALIZE (del_value);
+    }
+
+    property_p = next_prop;
```

Comment:
The code infitely loops for the following example:
```
Array.prototype.sort();
```

Reason is that `property_p` is not updated for internal properties.

It's better to put `property_p = next_prop` into `for`'s increment section.